### PR TITLE
Fro 13558, fix H2DataSourceMetaData.isInSameDatabaseInstance method, …

### LINKED ISSUE #13558: Add unit test for H2DataSourceMetaData.isInSameDatabaseInstance and DataSourceMetaData.isInSameDatabaseInstance
Hi community,

`H2DataSourceMetaData.isInSameDatabaseInstance` and `DataSourceMetaData.isInSameDatabaseInstance` was added recently, it need to add some unit tests for these methods, and verify the implementation of `isInSameDatabaseInstance`.

### PR DESCRIPTION
**Pull request by ReyYang**
Add unit test for H2DataSourceMetaData.isInSameDatabaseInstance and DataSourceMetaData.isInSameDatabaseInstance

Fixes #13558 .

Changes proposed in this pull request:
- fix H2DataSourceMetaData.isInSameDatabaseInstance method
- Add unit test for H2DataSourceMetaData.isInSameDatabaseInstance

### CODE REVIEW

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaData.java [hunk 1]
```diff
@@ -46,7 +50,9 @@
     private final Pattern pattern = Pattern.compile("jdbc:h2:((mem|~)[:/](?<catalog>[\\w\\-]+)|"
             + "(ssl:|tcp:)(//)?(?<hostName>[\\w\\-.]+)(:(?<port>[0-9]{1,4})/)?[/~\\w\\-.]+/(?<name>[\\-\\w]*)|"
             + "file:[/~\\w\\-]+/(?<fileName>[\\-\\w]*));?\\S*", Pattern.CASE_INSENSITIVE);
-    
+
+    private final Pattern modelPattern = Pattern.compile("((mem|~)[:/]|(ssl:|tcp:)|(file:))");
```
Comment (sandynz) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaData.java:54]: It's better to merge into one regex pattern for better performance.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaData.java [hunk 2]
```diff
@@ -46,7 +50,9 @@
     private final Pattern pattern = Pattern.compile("jdbc:h2:((mem|~)[:/](?<catalog>[\\w\\-]+)|"
```
Comment (sandynz) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaData.java:50]: `pattern` could be static for better performance. (Extra work)

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaData.java [hunk 3]
```diff
@@ -63,10 +69,23 @@ public H2DataSourceMetaData(final String url) {
...
     }
     
     @Override
     public boolean isInSameDatabaseInstance(final DataSourceMetaData dataSourceMetaData) {
+        if (!(dataSourceMetaData instanceof H2DataSourceMetaData)) {
+            return false;
+        }
+        if (!getModel().equals(((H2DataSourceMetaData) dataSourceMetaData).getModel())) {
+            return false;
+        }
```
Comment (sandynz) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaData.java:88]: Is `DataSourceMetaData.isInSameDatabaseInstance` necessary to update too?

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaData.java [hunk 4]
```diff
@@ -34,9 +34,13 @@
     private static final int DEFAULT_PORT = -1;
     
     private static final String DEFAULT_HOST_NAME = "";
-    
+
+    private static final String DEFAULT_H2_MODEL = "";
+
     private final String hostName;
-
```
Comment (sandynz) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaData.java:39]: Indents of empty lines should be kept.

File: shardingsphere-infra/shardingsphere-infra-common/src/test/java/org/apache/shardingsphere/infra/database/metadata/DataSourceMetaDataTest.java [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+        MySQLDataSourceMetaData actualMysql = new MySQLDataSourceMetaData("jdbc:mysql://127.0.0.1:9999/ds_0?serverTimezone=UTC&useSSL=false");
+        assertFalse(actualMysql.isInSameDatabaseInstance(actualH2));
+    }
+
+    @Test
+    public void assertFalseIsInSameDatabaseInstanceWithMysqlAndOracle() {
+        OracleDataSourceMetaData actualOracle = new OracleDataSourceMetaData("jdbc:oracle:thin:@//127.0.0.1:9988/ds_0", "test");
+        MySQLDataSourceMetaData actualMysql = new MySQLDataSourceMetaData("jdbc:mysql://127.0.0.1:9999/ds_0?serverTimezone=UTC&useSSL=false");
+        assertFalse(actualMysql.isInSameDatabaseInstance(actualOracle));
+    }
```
Comment (sandynz) [inline shardingsphere-infra/shardingsphere-infra-common/src/test/java/org/apache/shardingsphere/infra/database/metadata/DataSourceMetaDataTest.java:71]: We could simplify following similar test cases, just cover more code branches to improve test coverage.

File: shardingsphere-infra/shardingsphere-infra-common/src/test/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaDataTest.java [hunk 6]
```diff
@@ -68,4 +68,62 @@ public void assertNewConstructorWithFile() {
...
+        H2DataSourceMetaData actual1 = new H2DataSourceMetaData("jdbc:h2:~:ds-0;DB_CLOSE_DELAY=-1;DATABASE_TO_UPPER=false;MODE=MySQL");
+        H2DataSourceMetaData actual2 = new H2DataSourceMetaData("jdbc:h2:~:ds-1;DB_CLOSE_DELAY=-1;DATABASE_TO_UPPER=false;MODE=MySQL");
+        assertTrue(actual1.isInSameDatabaseInstance(actual2));
+    }
+
+    @Test
+    public void assertFalseIsInSameDatabaseInstance() {
+        H2DataSourceMetaData actual1 = new H2DataSourceMetaData("jdbc:h2:mem:ds_0;DB_CLOSE_DELAY=-1;DATABASE_TO_UPPER=false;MODE=MySQL");
+        H2DataSourceMetaData actual2 = new H2DataSourceMetaData("jdbc:h2:~:ds-1;DB_CLOSE_DELAY=-1;DATABASE_TO_UPPER=false;MODE=MySQL");
+        assertFalse(actual1.isInSameDatabaseInstance(actual2));
```
Comment (sandynz) [inline shardingsphere-infra/shardingsphere-infra-common/src/test/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaDataTest.java:90]: All of them are in memory, seems it should be true.

File: shardingsphere-infra/shardingsphere-infra-common/src/test/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaDataTest.java [hunk 7]
```diff
@@ -68,4 +68,62 @@ public void assertNewConstructorWithFile() {
...
+    }
+
+    @Test
+    public void assertFalseIsInSameDatabaseInstanceWithTcp() {
+        H2DataSourceMetaData actual1 = new H2DataSourceMetaData("jdbc:h2:tcp://localhost:8082/~/test1/test2;DB_CLOSE_DELAY=-1");
+        H2DataSourceMetaData actual2 = new H2DataSourceMetaData("jdbc:h2:tcp://192.168.64.76:8082/~/test3/test4;DB_CLOSE_DELAY=-1");
+        assertFalse(actual1.isInSameDatabaseInstance(actual2));
+    }
+
+
```
Comment (sandynz) [inline shardingsphere-infra/shardingsphere-infra-common/src/test/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaDataTest.java:107]: Just keep one empty line, delete duplicated ones. And also other places.

**Review by sandynz — CHANGES_REQUESTED** [review]

**Comment by ReyYang** [comment]
Roger that. I'll keep working on it

**Comment by ReyYang** [comment]
I have revised this issue as required, and I hope to get your opinion

File: shardingsphere-infra/shardingsphere-infra-common/src/test/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaDataTest.java [hunk 8]
```diff
@@ -20,8 +20,8 @@
 import org.junit.Test;
 
 import static org.hamcrest.CoreMatchers.is;
-import static org.junit.Assert.assertNull;
-import static org.junit.Assert.assertThat;
+import static org.junit.Assert.*;
```
Comment (sandynz) [inline shardingsphere-infra/shardingsphere-infra-common/src/test/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaDataTest.java:23]: Importing without star.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaData.java [hunk 9]
```diff
@@ -63,11 +71,35 @@ public H2DataSourceMetaData(final String url) {
...
     }
     
     @Override
     public boolean isInSameDatabaseInstance(final DataSourceMetaData dataSourceMetaData) {
+        if (!(dataSourceMetaData instanceof H2DataSourceMetaData)) {
+            return false;
+        }
+        if (!isSameModel(getModel(), ((H2DataSourceMetaData) dataSourceMetaData).getModel())) {
+            return false;
+        }
```
Comment (sandynz) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaData.java:91]: `mem`, `~`, `file`, `tcp-localhost`, they should be in the same instance.

**Review by sandynz — CHANGES_REQUESTED** [review]

File: shardingsphere-infra/shardingsphere-infra-common/src/test/java/org/apache/shardingsphere/infra/database/metadata/DataSourceMetaDataTest.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
+package org.apache.shardingsphere.infra.database.metadata;
```
Comment (sandynz) [inline shardingsphere-infra/shardingsphere-infra-common/src/test/java/org/apache/shardingsphere/infra/database/metadata/DataSourceMetaDataTest.java:1]: It need license header.

**Review by sandynz — CHANGES_REQUESTED** [review]

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaData.java (hunk 9 above)
Comment (sandynz) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/H2DataSourceMetaData.java:93]: Not include tcp

**Review by sandynz — APPROVED** [review]

**Pull request merged**